node/src/isolate/external.cpp

```diff
@@ -241,60 +254,75 @@ struct external_t::inner_t :
     }
 
     void connect() {
-        if(connecting) {
-            COCAINE_LOG_INFO(log, "connection to isolate daemon is already in progress");
+        if (std::chrono::system_clock::now() < (last_failed_connect_time + seal_time)) {
+            COCAINE_LOG_WARNING(log, "connection to isolate daemon is sealed");
             return;
         }
-        if(std::chrono::system_clock::now() < (last_failed_connect_time + seal_time)) {
-            COCAINE_LOG_WARNING(log, "connection to isolate daemon is sealed");
+
+        if (connecting.test_and_set()) {
+            COCAINE_LOG_INFO(log, "connection to isolate daemon is already in progress");
             return;
         }
-        connecting = true;
-        socket.reset(new asio::ip::tcp::socket(io_context));
-        if(session) {
+
+        if (session) {
             session->detach(std::error_code());
             session = nullptr;
         }
-        auto ep = args.as_object().at("external_isolation_endpoint", dynamic_t::empty_object).as_object();
+
+        const auto ep = args.as_object().at("external_isolation_endpoint", dynamic_t::empty_object).as_object();
         tcp::endpoint endpoint(
             asio::ip::address::from_string(ep.at("host", "127.0.0.1").as_string()),
             static_cast<unsigned short>(ep.at("port", 29042u).as_uint())
         );
 
-        auto connect_timeout_ms = args.as_object().at("connect_timeout_ms", 5000u).as_uint();
+        const auto connect_timeout_ms = args.as_object().at("connect_timeout_ms", 5000u).as_uint();
 
         COCAINE_LOG_INFO(log, "connecting to external isolation daemon to {}", boost::lexical_cast<std::string>(endpoint));
+
         auto self_shared = shared_from_this();
+        auto sentry = std::make_shared<flag_sentry_t>(connecting);
+        socket = std::make_unique<tcp::socket>(io_context);
 
         connect_timer.expires_from_now(boost::posix_time::milliseconds(connect_timeout_ms));
-        connect_timer.async_wait([=](const std::error_code& ec){
-            if(!ec) {
-                self_shared->socket->cancel();
```

Comment:
`self_shared` used to be captured, but now it's not. What guarantees that `this` won't be destroyed before this callback call?

Reply:
Yep, self_shared could extends inner_t lifetime beyond external_t lefitme, can't find out if it is desirable, but will fix to preserve original behavior.